### PR TITLE
feat: upgrade test

### PR DESCRIPTION
**Pull request by maan2003**
use `./scripts/upgrade-test.sh path/to/0.2/fedimintd` to run.

### CODE REVIEW

File: scripts/tests/upgrade-test.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# Runs a test to determine the latency of certain user actions
+
+set -euo pipefail
+export RUST_LOG="${RUST_LOG:-info}"
+
+source scripts/_common.sh
+build_workspace
+add_target_dir_to_path
+
+devimint upgrade-test "$@"
```
Comment (bradleystachurski) [inline scripts/tests/upgrade-test.sh:11]: Perhaps we can improve the usability of this script if we use `nix_build_binary_for_version` from `scripts/_common.sh` and pass versions to test as arguments.

```bash
versions=("$@")
for fed_version in "${versions[@]}"; do
  echo "Running upgrade test from fedimintd $fed_version to current"
  fedimintd_path="$(nix_build_binary_for_version 'fedimintd' "$fed_version")"
  export FM_BACKWARDS_COMPATIBILITY_TEST=1
  devimint upgrade-test "$fedimintd_path"
done
```

Then we can call with `./scripts/tests/upgrade-test.sh v0.2.1 v0.2.2` (and also add a `just` command).

I needed to add FM_BACKWARDS_COMPATIBILITY_TEST to handle the long latency for upgrading v0.2.1.

File: devimint/src/tests.rs [hunk 2]
```diff
@@ -2275,6 +2318,40 @@ pub async fn handle_command(cmd: TestCmd, common_args: CommonArgs) -> Result<()>
...
+            dev_fed.fed.start_server(&process_mgr, 3).await?;
+            info!("restarted waiting");
+            tokio::time::sleep(Duration::from_secs(10)).await;
+            info!("restarted waited");
+
+            stress_test_fed(&dev_fed).await?;
+            wait_session(&client).await?;
+            // kill 1 to check if 3 is working well
+            dev_fed.fed.terminate_server(1).await?;
+            tokio::try_join!(stress_test_fed(&dev_fed), wait_session(&client))?;
```
Comment (bradleystachurski) [inline devimint/src/tests.rs:2353]: [nit] A success log helps when parsing logs for multiple version upgrades

File: devimint/src/tests.rs [hunk 3]
```diff
@@ -2275,6 +2318,40 @@ pub async fn handle_command(cmd: TestCmd, common_args: CommonArgs) -> Result<()>
...
+            std::env::set_var("FM_FEDIMINTD_BASE_EXECUTABLE", old_fedimintd);
+            let mut dev_fed = dev_fed(&process_mgr).await?;
+            let client = dev_fed.fed.new_joined_client("test-client").await?;
+
+            tokio::try_join!(stress_test_fed(&dev_fed), wait_session(&client))?;
+
+            info!("shutting down servers");
+            dev_fed.fed.terminate_server(3).await?;
+            // kill 3rd peer early
+            tokio::time::sleep(Duration::from_secs(30)).await;
```
Comment (bradleystachurski) [inline devimint/src/tests.rs:2332]: Linter's failing, we can use `fedimint_core::task::sleep_in_test` instead of `tokio::time::sleep`.

**Review by bradleystachurski — COMMENTED** [review]
Thanks for putting this up!

I verified locally for v0.2.1 and v0.2.2 using the tweaked script in this review.

This tackles https://github.com/fedimint/fedimint/issues/4486 but we may want to also add tests for other binaries before closing.

**Comment by bradleystachurski** [comment]
We should also make a PR against master with these changes.

**Comment by elsirion** [comment]
This can be closed after https://github.com/fedimint/fedimint/pull/4762 landed I assume?

**Comment by bradleystachurski** [comment]
> This can be closed after #4762 landed I assume?

@elsirion Yep, thanks for closing 🙂

**Pull request closed without merging**